thorlcr/slave/traceslave.hpp

```diff
@@ -0,0 +1,203 @@
+
+#include "thmem.hpp"
+
+class CThorRowHistory : public OwnedConstThorRow
+{
+    int sequence;
+
+public:
+    inline CThorRowHistory() : sequence(0) {}
+
+    inline void set(const void * _row, unsigned long _sequence)
+    {
+        sequence = _sequence;
+        OwnedConstThorRow::set(_row);
+    }
+    inline int getSequence() const { return sequence;}
+};
+
+class CTracingThorDataLink : implements CInterfaceOf<IThorDataLink>
+{
+private:
+    class TraceQueue
+    {
+    public:
+        CThorRowHistory *rowBuffer;
+        unsigned rowBufferSize;
+        unsigned rowTailp;
+
+        TraceQueue() : rowBuffer(NULL), rowBufferSize(0), rowTailp(0){ }
+
+        ~TraceQueue()
+        {
+            delete [] rowBuffer;
+        }
+        void init(unsigned _bufsize)
+        {
+            if (!rowBuffer && _bufsize)
+            {
+                rowBufferSize = _bufsize;
+                rowBuffer = new CThorRowHistory[rowBufferSize];
+                rowTailp = 0;
+            }
+        }
+        void enqueue(const void *row, int sequence)
+        {
+            // NOTE - updates then sets rowHeadp (atomically) so that the code
+            // below (ignoring oldest record) is safe
+            rowBuffer[rowTailp].set(row,sequence);
+            unsigned nextTailp = rowTailp + 1;
+            if (nextTailp==rowBufferSize)
+                nextTailp = 0;
+            rowTailp = nextTailp;
+        }
+        // Move items in current queue into a second queue
+        // clearing original queue
+        void queueOut(TraceQueue &traceOutQueue)
+        {
+            unsigned rowPos = rowTailp + 1; // this is the tail
+            if (rowPos==rowBufferSize) rowPos = 0;
+            // NOTE - we don't output every item in the list as one
+            // COULD have been updated after we swapped the buffers
+            // - queue out all but last item
+            for (unsigned n=1; n<rowBufferSize;++n)
+            {
+                if (rowBuffer[rowPos])
+                {
+                  traceOutQueue.enqueue(rowBuffer[rowPos].get(),rowBuffer[rowPos].getSequence());
+                  rowBuffer[rowPos].clear();
+                }
+                if (++rowPos==rowBufferSize) rowPos = 0; // NOTE - skipping first item intentional
+            }
+        }
+        // Dump items to MemoryBuffer (without clearing items form queue)
+        void dump(MemoryBuffer &mb, IHThorArg *helper) const
+        {
+            unsigned rowPos = rowTailp+1; // this is the tail
+            if (rowPos==rowBufferSize) rowPos = 0;
+
+            CommonXmlWriter xmlwrite(XWFnoindent|XWFtrim);
+            IOutputMetaData *meta = helper->queryOutputMeta();
+            for (unsigned n=0; n<rowBufferSize;++n)
+            {
+                if (rowBuffer[rowPos])
+                {
+                    xmlwrite.outputBeginNested("Row", true);
+                    xmlwrite.outputInt(rowBuffer[rowPos].getSequence(), sizeof(int), "@seq");
```

Comment:
Sequence is provided in HThor and Roxie's debugger response.  I assume that there is a reason for providing this to ECLIDE?